### PR TITLE
Support YYYY-MM-DD (ISO-8601) in observation files, timemap files and END_DATE keyword.

### LINKED ISSUE #2639: YYYY-MM-DD (ISO-8601) support in ERT configuration
**Is your feature request related to a problem? Please describe.**
ERT configuration does not support dates written in ISO-8601 (YYYY-MM-DD) in configuration files and observation files.

**Describe the solution you'd like**
YYYY-MM-DD is recognized automatically alongside MM/DD/YYYY.  There is already support for iso-dates in `ecl`, in `util_sscanf_isodate`.  The function `util_sscanf_date_utc` is currently in use in ERT, and returns ok, but a negative `time_t` if given an ISO-formatted date (!). 

Supporting ISO-dates early even for ERT2 can be useful in order to be able to deprecate MM/DD/YYYY later, getting users to fix configuration files requires a long deprecation period.

**Describe alternatives you've considered**
Await ERT3 yaml configuration format for ERT, where support for ISO-8601 can be exclusive.

**Additional context**
https://equinor.slack.com/archives/CDPJULWR4/p1640196273105900

### PR DESCRIPTION
**Pull request by berland**
**Issue**
Resolves #2639

**Approach**
ISO-date-parsing is already supported in `ecl`. Add support for both ISO-8601 (YYYY-MM-DD) and legacy DD/MM/YYYY  by always trying to parse with ISO first, and if it fails, try with DD/MM/YYYY. 

Note 1: The legacy date format accepts any non-decimal character as the separator.
Note 2: The legacy date format accepts ISO-strings as dates, but will give an unexpected date as a result (1997-06-07 ends in somewhere in  December year 0018), this means we must test for ISO validity first.

## Pre review checklist

- [x] Added appropriate labels

### CODE REVIEW

File: libres/lib/enkf/ecl_config.cpp [hunk 1]
```diff
@@ -367,12 +367,16 @@ static void handle_has_end_date_key(ecl_config_type *ecl_config,
                                     const config_content_type *config) {
     const char *date_string = config_content_get_value(config, END_DATE_KEY);
     time_t end_date;
-    if (util_sscanf_date_utc(date_string, &end_date))
+    bool end_date_parsed_ok;
+    end_date_parsed_ok = util_sscanf_isodate(date_string, &end_date);
```
Comment (markusdregi) [inline libres/lib/enkf/ecl_config.cpp:371]: Any reason to split declaration and definition over two lines?

File: libres/lib/enkf/time_map.cpp [hunk 2]
```diff
@@ -152,7 +152,12 @@ bool time_map_fscanf(time_map_type *map, const char *filename) {
                 char date_string[128];
                 if (fscanf(stream, "%s", date_string) == 1) {
                     time_t date;
-                    if (util_sscanf_date_utc(date_string, &date)) {
+                    bool date_parsed_ok;
+                    date_parsed_ok = util_sscanf_isodate(date_string, &date);
```
Comment (markusdregi) [inline libres/lib/enkf/time_map.cpp:156]: Same question regarding declaration here...

File: res/enkf/ecl_config.py [hunk 3]
```diff
@@ -111,11 +111,18 @@ def __init__(self, config_content=None, config_dict=None):
                 refcase_list.append(refcase)
 
             # END_DATE_KEY
-            end_date = CTime(
-                datetime.strptime(
-                    config_dict.get(ConfigKeys.END_DATE, "31/12/1969"), "%d/%m/%Y"
+            try:
+                end_date = CTime(
+                    datetime.strptime(
+                        config_dict.get(ConfigKeys.END_DATE, "31/12/1969"), "%d/%m/%Y"
```
Comment (markusdregi) [inline res/enkf/ecl_config.py:117]: Any reason we are not here as well trying to first load using ISO-format?

**Review by markusdregi — COMMENTED** [review]
I think this PR looks really good 👍🏻 Left a few minor comments...

A last thing we should discuss is whether we want to deprecate the current date format or not. Because if we want to, we should either add a deprecation warning or create an issue on doing so...

File: libres/lib/enkf/ecl_config.cpp (hunk 1 above)
Comment (berland) [reply libres/lib/enkf/ecl_config.cpp:371]: Nope.. fixed.

File: libres/lib/enkf/time_map.cpp (hunk 2 above)
Comment (berland) [reply libres/lib/enkf/time_map.cpp:156]: fixed.

File: res/enkf/ecl_config.py (hunk 3 above)
Comment (berland) [reply res/enkf/ecl_config.py]: No, both orders work here. Changed.

**Comment by berland** [comment]
A word of warning: This PR removes support for DD-MM-YYYY which at least parts of ERT supports pr. today. This date format has never been advertised as working, and not mentioned in documentation. But if someone is accidentally using it today, this PR will make it crash for them (ERT  (ecl)  happily supports any non-decimal separator in DDxMMxYYYY).

**Comment by markusdregi** [comment]
> A word of warning: This PR removes support for DD-MM-YYYY which at least parts of ERT supports pr. today. This date format has never been advertised as working, and not mentioned in documentation. But if someone is accidentally using it today, this PR will make it crash for them (ERT (ecl) happily supports any non-decimal separator in DDxMMxYYYY).

I think that is fair! The error message is self-explanatory, it makes sense and it is requested that we support ISO dates and the above format was never documented to be supported anyway... It might be that some users will have to update their setups, but I think it is worth it 👍🏻 

Will you clean up your commit history and make the title to your PR have enough context to serve as a line in the release notes? Besides that I consider this PR ready for merging 🚀

**Comment by berland** [comment]
> I think this PR looks really good 👍🏻 Left a few minor comments...
> 
> A last thing we should discuss is whether we want to deprecate the current date format or not. Because if we want to, we should either add a deprecation warning or create an issue on doing so...

See #2655

**Comment by berland** [comment]
The CI failure reason is flakiness, and is the same as in #2429

**Comment by markusdregi** [comment]
Jenkins: Retest this please

**Review by markusdregi — APPROVED** [review]
🚀

**Pull request merged**